File: learn/src/main/java/org/platanios/learn/graph/HITSAlgorithm.java
package org.platanios.learn.graph;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

/**
 * @author Emmanouil Antonios Platanios
 */
public class HITSAlgorithm<E> {
    private final Logger logger = LogManager.getFormatterLogger("HITS");

    private final Graph<VertexContentType, E> graph;
    private final int maximumNumberOfIterations;
    private final boolean checkForScoresConvergence;
    private final double scoresChangeTolerance;
    private final int loggingLevel;

    private int currentIteration = 0;
    private double lastMeanAbsoluteRankChange = Double.MAX_VALUE;
    private double authorityNormalizationConstant = 0.0;
    private double hubNormalizationConstant = 0.0;

    public static class Builder<E> {
        private final Graph<VertexContentType, E> graph;

        private int maximumNumberOfIterations = 30;
        private boolean checkForScoresConvergence = true;
        private double scoresChangeTolerance = 1e-6;
        private int loggingLevel = 0;

        public Builder(Graph<VertexContentType, E> graph) {
            this.graph = graph;
        }

        public Builder maximumNumberOfIterations(int maximumIterations) {
            this.maximumNumberOfIterations = maximumIterations;
            return this;
        }

        public Builder checkForScoresConvergence(boolean checkForScoresConvergence) {
            this.checkForScoresConvergence = checkForScoresConvergence;
            return this;
        }

        public Builder scoresChangeTolerance(double scoresChangeTolerance) {
            this.scoresChangeTolerance = scoresChangeTolerance;
            return this;
        }

        public Builder loggingLevel(int loggingLevel) {
            this.loggingLevel = loggingLevel;
            return this;
        }

        public HITSAlgorithm<E> build() {
            return new HITSAlgorithm<>(this);
        }
    }

    private HITSAlgorithm(Builder<E> builder) {
        graph = builder.graph;
        maximumNumberOfIterations = builder.maximumNumberOfIterations;
        checkForScoresConvergence = builder.checkForScoresConvergence;
        scoresChangeTolerance = builder.scoresChangeTolerance;
        loggingLevel = builder.loggingLevel;
    }

    public void computeScores() {
        while (!checkTerminationConditions()) {
            performIterationUpdates();
            currentIteration++;
            if ((loggingLevel == 1 && currentIteration % 1000 == 0)
                    || (loggingLevel == 2 && currentIteration % 100 == 0)
                    || (loggingLevel == 3 && currentIteration % 10 == 0)
                    || loggingLevel > 3)
                logIteration();
        }
    }

    private boolean checkTerminationConditions() {
        return currentIteration >= maximumNumberOfIterations
                || checkForScoresConvergence && lastMeanAbsoluteRankChange <= scoresChangeTolerance;
    }

    // TODO: Can be made more efficient by calling the update vertices content method only once.
    private void performIterationUpdates() {
        graph.computeVerticesUpdatedContent(this::vertexComputeFunction);
        graph.updateVerticesContent();
        authorityNormalizationConstant = Math.sqrt(authorityNormalizationConstant);
        hubNormalizationConstant = Math.sqrt(hubNormalizationConstant);
        graph.computeVerticesUpdatedContent(this::normalizationVertexComputeFunction);
        graph.updateVerticesContent();
        authorityNormalizationConstant = 0.0;
        hubNormalizationConstant = 0.0;
        if (checkForScoresConvergence) {
            lastMeanAbsoluteRankChange = 0.0;
            for (Vertex<VertexContentType, E> vertex : graph.getVertices())
                lastMeanAbsoluteRankChange += vertex.getContent().lastAbsoluteChange;
            lastMeanAbsoluteRankChange /= graph.getNumberOfVertices();
        }
    }

    private void logIteration() {
        if (checkForScoresConvergence)
            logger.info("Iteration #: %10d | Mean Absolute Rank Change: %20s", currentIteration, lastMeanAbsoluteRankChange);
        else
            logger.info("Iteration #: %10d", currentIteration);
    }

    private VertexContentType vertexComputeFunction(Vertex<VertexContentType, E> vertex) {
        double oldAuthorityScore = vertex.getContent().authorityScore;
        double oldHubScore = vertex.getContent().hubScore;
        double newAuthorityScore = 0.0;
        double newHubScore = 0.0;
        for (Edge<VertexContentType, E> incomingEdge : vertex.getIncomingEdges())
            newAuthorityScore += incomingEdge.getSourceVertex().getContent().hubScore;
        for (Edge<VertexContentType, E> outgoingEdge : vertex.getOutgoingEdges())
            newHubScore += outgoingEdge.getDestinationVertex().getContent().authorityScore;
        authorityNormalizationConstant += newAuthorityScore * newAuthorityScore;
        hubNormalizationConstant += newHubScore * newHubScore;
        return new VertexContentType(vertex.getContent().id,
                                     newAuthorityScore,
                                     newHubScore,
                                     checkForScoresConvergence ?
                                             Math.abs(newAuthorityScore - oldAuthorityScore)
                                                     + Math.abs(newHubScore - oldHubScore)
                                             : 0.0);
    }

    private VertexContentType normalizationVertexComputeFunction(Vertex<VertexContentType, E> vertex) {
        return new VertexContentType(vertex.getContent().id,
                                     vertex.getContent().authorityScore / authorityNormalizationConstant,
                                     vertex.getContent().hubScore / hubNormalizationConstant,
                                     vertex.getContent().lastAbsoluteChange);
    }

    public static class VertexContentType {
        private final int id;
        private final double authorityScore;
        private final double hubScore;
        private final double lastAbsoluteChange;

        public VertexContentType(int id, double authorityScore, double hubScore) {
            this(id, authorityScore, hubScore, Double.MAX_VALUE);
        }

        public VertexContentType(int id, double authorityScore, double hubScore, double lastAbsoluteChange) {
            this.id = id;
            this.authorityScore = authorityScore;
            this.hubScore = hubScore;
            this.lastAbsoluteChange = lastAbsoluteChange;
        }

        public int getId() {
            return id;
        }

        public double getAuthorityScore() {
            return authorityScore;
        }

        public double getHubScore() {
            return hubScore;
        }

        public double getLastAbsoluteChange() {
            return lastAbsoluteChange;
        }
    }
}

File: learn/src/main/scala/makina/scala/learn/graph/HITSAlgorithm.scala
package makina.scala.learn.graph

import com.typesafe.scalalogging.Logger
import makina.scala.learn.graph.HITSAlgorithm._
import org.slf4j.LoggerFactory

/**
  * @author Emmanouil Antonios Platanios
  */
case class HITSAlgorithm[E >: Null <: AnyRef](
  graph: Graph[VertexContent, E],
  maximumNumberOfIterations: Int = 30,
  checkForScoresConvergence: Boolean = true,
  scoresChangeTolerance: Double = 1e-6,
  loggingLevel: Int = 0
) {
  private var currentIteration: Int = 0
  private var lastMeanAbsoluteScoresChange: Double = Double.MaxValue
  private var authorityNormalizationConstant: Double = 0.0
  private var hubNormalizationConstant: Double = 0.0

  def computeRanks() = {
    while (!checkTerminationConditions()) {
      performIterationUpdates()
      currentIteration += 1
      if ((loggingLevel == 1 && currentIteration % 1000 == 0) ||
          (loggingLevel == 2 && currentIteration % 100 == 0) ||
          (loggingLevel == 3 && currentIteration % 10 == 0) ||
          loggingLevel > 3)
        logIteration()
    }
  }

  private def checkTerminationConditions(): Boolean = {
    currentIteration >= maximumNumberOfIterations ||
    (checkForScoresConvergence && lastMeanAbsoluteScoresChange <= scoresChangeTolerance)
  }

  // TODO: Can be made more efficient by calling the update vertices content method only once.
  private def performIterationUpdates() = {
    graph.computeVerticesUpdatedContent(vertexUpdateFunction)
    graph.updateVerticesContent()
    authorityNormalizationConstant = Math.sqrt(authorityNormalizationConstant)
    hubNormalizationConstant = Math.sqrt(hubNormalizationConstant)
    graph.computeVerticesUpdatedContent(normalizationVertexUpdateFunction)
    graph.updateVerticesContent()
    authorityNormalizationConstant = 0.0
    hubNormalizationConstant = 0.0
    if (checkForScoresConvergence) {
      lastMeanAbsoluteScoresChange = 0.0
      for (vertex <- graph.vertices)
        lastMeanAbsoluteScoresChange += vertex.content.lastAbsoluteChange
      lastMeanAbsoluteScoresChange /= graph.numberOfVertices
    }
  }

  private def logIteration() = {
    if (checkForScoresConvergence)
      logger.info(f"Iteration #: $currentIteration%10d | " +
                  f"Mean Absolute Scores Change: $lastMeanAbsoluteScoresChange%20s")
    else
      logger.info(f"Iteration #: $currentIteration%10d")
  }

  private def vertexUpdateFunction(vertex: Vertex[VertexContent, E]): VertexContent = {
    val oldAuthorityScore: Double = vertex.content.authorityScore
    val oldHubScore: Double = vertex.content.hubScore
    var newAuthorityScore: Double = 0.0
    var newHubScore: Double = 0.0
    vertex.incomingEdges.foreach(newAuthorityScore += _.source.content.hubScore)
    vertex.outgoingEdges.foreach(newHubScore += _.destination.content.authorityScore)
    authorityNormalizationConstant += newAuthorityScore * newAuthorityScore
    hubNormalizationConstant += newHubScore * newHubScore
    VertexContent(
      vertex.content.id,
      newAuthorityScore,
      newHubScore,
      if (checkForScoresConvergence)
        Math.abs(newAuthorityScore - oldAuthorityScore) + Math.abs(newHubScore - oldHubScore)
      else
        0.0
    )
  }

  private def normalizationVertexUpdateFunction(vertex: Vertex[VertexContent, E]): VertexContent = {
    VertexContent(
      vertex.content.id,
      vertex.content.authorityScore / authorityNormalizationConstant,
      vertex.content.hubScore / hubNormalizationConstant,
      vertex.content.lastAbsoluteChange
    )
  }
}

object HITSAlgorithm {
  val logger = Logger(LoggerFactory.getLogger("Graph / HITS"))

  case class VertexContent(id: Int,
                           authorityScore: Double,
                           hubScore: Double,
                           lastAbsoluteChange: Double = Double.MaxValue)
}
